rust-bins/src/bin/keygen.rs

```diff
@@ -301,6 +468,220 @@ pub fn generate_ed_sk(
+    let hash = Sha256::digest(&bit_vec.into_vec());
+
+    // convert hash from byte vector to bit vector
+    let hash_bits = BitVec::<Msb0, u8>::from_slice(&hash).unwrap();
+
+    // sentence is valid if checksum equals fist ent_len/32 bits of hash
+    checksum == hash_bits[0..ent_len / 32]
+}
+
+/// Convert given byte array to valid BIP39 sentence.
+/// Bytes must contain {16, 20, 24, 28, 32} bytes corresponding to
+/// {128, 160, 192, 224, 256} bits.
+/// This uses the method described at https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki
+pub fn bytes_to_bip39(bytes: &[u8], bip_word_list: &[&str]) -> Result<Vec<String>, String> {
+    let ent_len = 8 * bytes.len(); // input is called entropy in BIP39
+    match ent_len {
+        128 | 160 | 192 | 224 | 256 => (),
+        _ => {
+            return Err(
+                "The number of bytes to be converted to a BIP39 sentence must be in {16, 20, 24, \
+                 28, 32}."
+                    .to_string(),
+            )
+        }
+    };
+
+    // checksum length is ent_len / 32
+    let cs_len = ent_len / 32;
+
+    // checksum is first cs_len bits of SHA256(bytes)
+    // first compute hash of bytes
+    let hash = Sha256::digest(bytes);
+
+    // convert hash from byte vector to bit vector
+    let hash_bits = succeed_or_die!(
+        BitVec::<Msb0, u8>::from_slice(&hash),
+        e => "Failed to convert hash to bit vector because {}"
+    );
+
+    // convert input bytes from byte vector to bit vector
+    let mut random_bits = succeed_or_die!(
+        BitVec::<Msb0, u8>::from_slice(&bytes),
+        e => "Failed to convert hash to bit vector because {}"
+    );
+
+    // append the first cs_len bits of hash_bits to the end of random_bits
+    for i in 0..cs_len {
+        random_bits.push(hash_bits[i]);
+    }
+
+    // go over random_bits in chunks of 11 bits and convert those to words
+    let mut vec = Vec::<String>::new();
+    let random_iter = random_bits.chunks(11);
+    for chunk in random_iter {
+        let idx = chunk.iter().fold(0, |acc, b| acc << 1 | *b as usize); // convert chunk to integer
+        vec.push(bip_word_list[idx].to_string());
+    }
+
+    Ok(vec)
+}
+
+/// Rerandomize given list of words using system randomness and HKDF extractor.
+/// The input can be an arbitrary slice of strings.
+/// The output is a valid BIP39 sentence with 24 words.
+pub fn rerandomize_bip39(
+    input_words: &[String],
+    bip_word_list: &[&str],
+) -> Result<Vec<String>, String> {
+    // Get randomness from system.
+    // Fill array with 256 random bytes, corresponding to 2048 bits.
+    let mut system_randomness = [0u8; 256];
+    rand::thread_rng().fill(&mut system_randomness[..]);
+
+    // Combine both sources of randomness using HKDF extractor.
+    // Using salt with randomness from random.org for added security.
+    let salt = b"keygen-rand-wEtIBpTIyzPRpZxNUIherQh14uPlDIdiqngFSo1qrqE1UrXl5DcUfV4xddYNDnOMIumlkqS9HNshATaFxAwqiUtLj5rxeBJIOsav";
```

Comment:
I think this is a bit dodgy since this is unverifiable that you have done so.

Would it not make more sense to have something plain and boring here since the value is public, e.g., 

`concordium-key-generation-tool-version-1`